[RFC] Batch processor migration (#15273)
open-telemetry/opentelemetry-collector: docs/rfcs/batching-migration.md

# Migration for `batchprocessor` and exporterhelper batching

## Overview

The Collector now offers two batching mechanisms:

1. The `batchprocessor` component, used in pipelines for many years and
   widely deployed, notably as a default in the OpenTelemetry Helm
   chart and other distributions.
2. The `exporterhelper` combined queue/batch sender, a replacement for
   the batch processor, is ready to replace it.

> We find there are a small number of users and known use-cases which

Reviewer: are the use-cases documented, would be great to have them in here

Author: Added the ones we know.

error propagation. With only one thread exporting batches, performance
suffers unless the downstream component acts the same. This explains
why the exporterhelper's `wait_for_result` queue sender setting must
remain off by default. Batch processor stands in the way of error
propagation with concurrency, so most users do not propagate errors.

### Exporterhelper replacement

The `exporterhelper` queue/batch sender resolves the known
batchprocessor defects and integrates new features:

- Choice of sizing logic by items, bytes, and requests
- Optional synchronous or asynchronous behavior
- Optional persistent storage extension
- Optional blocking

At this time, the `exporterhelper` queue/batch sender has feature
parity with `batchprocessor`. However, its default settings are still
aligned with the batch processor:

- `wait_for_result: false` we do not propagate errors
- `batch::enabled: false` we disable exporter batching

The exporterhelper supports opt-in persistent storage, but the default
in-memory queue combined with `wait_for_result: false` establishes
a potential to drop data while suppressing error propagation.

The exporterhelper's built-in batching support is critical for

> vendors, because it gives exporters a way to control request size in

Reviewer 2: ```suggestion
vendor-specific components, because it gives exporters a way to control request size in
```

terms of bytes, considering their own protocol. The ideal final state
is that we enable exporterhelper batching by default. Many exporters
already do this, but not all, the choice is made per-component.

> #### Recommendation: error propagation if no persistent store

Reviewer 3: Is this recommendation for exporter as well, or only for the batch processor "re-designed"?

Reviewer 4: I'm concerned here as well. We are effectively proposing a change to the default queue behavior to make it synchronous. This should be out of scope for the batch migration.

Author: I guess this is the real debate here. I believe users are not well served by out-of-the-box defaults that allow the collector to drop data after upstream failure(s) while returning success to the caller. My statement is emphatically meant to apply for OpenTelemetry SDKs sending to OTLP receiver through to OTLP exporter. The OTel SDKs would not specify to use batching processors by default and implement their own dropping in an environment where collectors drop-suppress both permanent and transient errors and disable backpressure.

I understand that @dashpole has raised compatibility concerns with PRW's handling of single-writer concepts, how it does not tolerate out-of-order arrival. To me, this suggests that PRW should not include the QueueBatch helper; OTOH, I would also expect correct behavior for a PRW exporter configured with a queue and batch settings (to allow splitting) that used a single num_consumers=1, which I expect to preserve ordering.

You can convince me to agree that a change to synchronous-by-default is out of scope for the batch processor migration specifically, but I still think dropping-by-default is a 1.0 release blocker of an issue.

Author: I reversed my position on wait_for_result, and have documented more of the reasons why some exporters will opt-out.

We aim to set `wait_for_result: true` when there is no storage
extension configured. This flag is already only applicable when the

> in-memory queue is used, storage extensions already ignore this mode.

Reviewer 5: Bit of an aside, but perhaps we should make it an error to specify `wait_for_result: true` with a storage extension

Author: I added an additional work item.

A feature flag for this change will be defined,
`exporterhelper.memoryQueueWaitForResult`.

#### Recommendation: exporterhelper batching enabled by default

Currently, components have a mixture of `configoptional.Some`,
`configoptional.Default`, and `configoptional.None`, usually passing
`exporterhelper.DefaultQueueBatchConfig()` which is defined:

```go
func NewDefaultQueueConfig() queuebatch.Config {
	return queuebatch.Config{
		Sizer:           request.SizerTypeRequests,
		NumConsumers:    10,
		QueueSize:       1_000,
		BlockOnOverflow: false,
        WaitForResult:   false,
		Batch: configoptional.Default(queuebatch.BatchConfig{
			FlushTimeout: 200 * time.Millisecond,
			Sizer:        request.SizerTypeItems,
			MinSize:      8192,
		}),
	}
}
```

The `configoptional.Default` is responsible for disabling batching for
most users. To complete the migration, in the final state, we will have
a single standard constructor where `wait_for_result` is true and batch
is enabled by default:

```go

> func StandardQueueConfig() configoptional.Optional[queuebatch.Config] {

Reviewer 5: ```suggestion
func StandardQueueConfig() queuebatch.Config {
```

I think the caller would be responsible for wrapping it in a `configoptional.Default`?

> func StandardQueueConfig() configoptional.Optional[queuebatch.Config] {

Reviewer 6: I'm not sure we need a separate function, especially considering that this contradicts our naming guidelines (["Methods that return a variable that uses non-zero value(s) that impacts business logic MUST use the prefix NewDefault"](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/coding-guidelines.md#go-api-naming-conventions)). The existing function can change its return value depending on the state of the feature gate, and components that don't want the new default should override it in their `createDefaultConfig` callback. The hard part is ensuring that component authors are made aware of the change before the feature gate goes beta, so they're not caught off guard by new batching logic they might not want or need.

Author: I simplified this proposal. It was too complex, and this has been removed.

	return queuebatch.Config{
        // ...
        WaitForResult:   true,
		Batch: configoptional.Some(queuebatch.BatchConfig{
            // ...
		}),
	}
}
```

The implementation will be determined by the feature flags discussed
in this document. There are two levels of `configoptional` here, one
for the queue and one for the batcher: both will become enabled by
default at the end of this migration.

A feature flag for this change will be defined,
`exporterhelper.exporterQueueBatchEnabled`.

### Double-batching problem

One concern preventing the migration is that we may unintentionally
apply multiple batching processes in a single pipeline. This is a
coordination problem. If a pipeline with a default- or user-configured
batch processor has settings that are out of line with the new
exporterhelper defaults, users will pay the cost of batching and then

> re-batching.

Reviewer 3: the other problem is this, in a pipeline "batch -> new_exporter_with_batching_and_wait_for_result" because batch is single thread, you essentially limit the throughput to 1 request per flush interval of the exporter batch, so a significant hit in performance and throughput.

Author: The plan that I am outlining would make the batch processor into a No-op at the same instant the default batching is enabled. I mean for this instantaneous switch to address your concern.

> #### Recommendation: coordinated batch processor phase-out

Reviewer 3: There are some exporters that "because exporter helper sending queue was not stable" explicitly disable sending_queue and batching. We should revisit all of the exporters in contrib if not already done that.

Author: Yes. This is what I mean when I called for an audit ("At the same time, an audit is required to clean up the many various initial conditions across exporters.")

> Using a single feature flag, we will enable exporterhelper

Reviewer: I wonder if this will work for end users, having the batching be moved from processors to exporters once the flag is enabled by default could cause unexpected problems. I guess the transition period could be long enough between introducing the new processor, and the time when this flag is enabled to give users ample time to transition.

I wonder if the time when this flag is enabled by default (batching at the exporter) should also be the time when the batch processor is removed, to prevent someone in a complex case from not being aware of the underlying change.

Author: To me, the default behavior is an unexpected problem that we are trying to fix.

My intention was to declare the deprecation period _without changing exporterhelper defaults_ for six months, giving users that long to install a proper batch processor and disable exporterhelper settings. Removing the batch processor from the release is appropriate, maybe a little aggressive. I would agree to that.

Author: I revised the plan outline following your guidance.  The batchprocessor will be removed from the core/contrib distros before changing the default exporterhelper.

queue/batching by default, at the same time we disable the batch

> processor by rendering it a no-op. First, we will deprecate the batch

Reviewer 6: I think making the batch processor a no-op through a feature gate is a bad idea. Users should be incentivized to remove it themselves, especially since having double batching seems to me like much less of a problem than the possibility of suddenly having no batching at all, for instance because you're using a custom exporter which doesn't use exporterhelper and didn't hear about the migration.

Reviewer 7: Would it be feasible to disallow pipelines with both a batch processor and the batch exporter helper enabled in the same pipeline (i.e. treat it as invalid during component initialization)?  Are there known use-cases where a user needs both? Then the user can remediate by removing one or the other.

Reviewer 6: If some processors have efficiency concerns / payload size restrictions like exporters often do, I could see a use case for batching twice with different settings. If we detect such a pipeline, I would suggest _at most_ logging a warning about inefficiency.

Author: The current batch processor is fundamentally broken, that's why we're removing it. If users keep the old batch processor alive, they have the defects we know about: no concurrency, no error propagation, and no tracing propagation. The period of six months defined in this plan is to let the migrate before anything breaks (i.e., they should "remove it themselves" within six months).

Reviewer 6: (Duplicating what I said on Slack for visibility) I agree that they need to migrate, the problem is that making the batch processor a no-op at the end of the migration period is just breaking the behavior further, instead of giving a clear signal that they need to upgrade because we no longer support the processor. To me, a clearer signal would be (for instance) an error message at startup because the processor no longer exists.

Author: I've removed the "magic" from this proposal. The batchprocessor will be removed before changing the defaults according to the new plan.

processor: users will be guided in two directions:

1. In simple pipelines, the user will remove the batch processor from
   their configuration and configure the queue_sender::batch instead.
2. In complex pipelines, a new core processor will be introduced as a
   from-scratch replacement, called the `queuebatchprocessor`. The new
   core processor will be implemented by exporterhelper internally, as
   demonstrated in
   [#13583](https://github.com/open-telemetry/opentelemetry-collector/pull/13583). This
   component will support configuration identical with `queue_sender`
   but for use as a processor.

At the same time, an audit is required to clean up the many various
initial conditions across exporters. We will define an Go package
`batchmigration` with an enum to capture all the common initial

> settings,

Reviewer 6: I'm not sure I understand what exactly "cleaning up initial conditions" means, or what the motivation behind it is, could you elaborate? I'm especially not sure I understand the point of the `batchmigration` package.

I think it's fine, or even expected, for different exporters to have different default batching settings. If what we want is just to _enable_ batching by default to reduce the config changes needed to remove the batch processor, and we think the worst case scenario is just excessive batching, I would suggest a simpler migration:
- Changing the output of `NewDefaultQueueConfig` to enable batching by default under an Alpha feature gate
- Auditing exporters in Contrib with code owners to determine if they have reasons to deviate from standard practice (queue + batching enabled). If yes, they can keep omitting the `queue_sender` field, or they can override the default config to disable batching by default; if not, we make sure they have `queue_sender` in their config.
- Once we are confident that components that don't want the new upstream default have updated to override it, advance the feature gate to beta.

Reviewer 2: Agreed with this, I think it is okay that there are different configurations per exporters. For example, pull-based exporters may don't care much about batching or queueing

Author: Any exporter that does not want queue/batch behavior can simply not install the QueueBatch sender in their configuration or directly configure hard-coded queue/batch settings in their constructor. Do you agree that special cases can simply bypass the feature they do not want?

Cleaning up initial conditions are what Bogdan referred to in https://github.com/open-telemetry/opentelemetry-collector/pull/15273/changes/BASE..6beccb809ee7a80430a54165092672e50be36020#r3213954079, the users who disabled batching simply because it wasn't ready.

Reviewer 2: > Do you agree that special cases can simply bypass the feature they do not want?

Yes, that makes sense

> Cleaning up initial conditions are what Bogdan referred to in https://github.com/open-telemetry/opentelemetry-collector/pull/15273/changes/BASE..6beccb809ee7a80430a54165092672e50be36020#r3213954079, the users who disabled batching simply because it wasn't ready.

Alright, I think it would be easier to understand as "unify the settings for exporters that may have disabled this because of stability concerns"?

Author: I considered whether we should have multiple defaults, potentially a shared default for exporters that want to be (a) synchronous, (b) single concurrency, etc.

```go
// This will be deprecated and removed after the migration.
type Existing int

const (
    // Default means the caller used configoptional.Default(NewDefaultQueueConfig())
    Default Case = iota

>     // Default means the caller used configoptional.Some(NewDefaultQueueConfig())

Reviewer 8: nit - typo:
```suggestion
    // Some means the caller used configoptional.Some(NewDefaultQueueConfig())
```

Author: Sorry I rewrote this entirely, I appreciate your review. I believe I've corrected this sort of mistake--the new proposal is a lot simpler, this was removed.

    Some Case

>     // Default means the caller used configoptional.None[...]()

Reviewer 8: nit - typo:
```suggestion
    // None means the caller used configoptional.None[...]()
```

    None Case

    // The callsite enables queue, not batch.
    SomeQueueNoneBatch,

    // ...
)
```

Exporters will be able opt-out of the standard migration path by
rejecting the proposed changes during an audit step, where most callers
will be migrated using a new function,

```go
func MigrateQueueConfig(migrate batchmigration.Existing) configoptional.Optional[queuebatch.Config] {
    // Each case will reproduce exactly the behavior of a callsite
    // somewhere in core or collector-contrib.
}
```

A number of exporters today have custom settings because the default
settings today cause data loss. After the migration is complete, these

> components should prefer to switch to standard queue config. For these

Reviewer 7: Just to enumerate a few reasons, in-case this is helpful:

* Prometheus exporter has no need for queueing or batching because it is pull
* Prometheus remote write exporter can't use the queue because it doesn't guarantee points are delivered in-order
* Google cloud exporter disables retries because those are already handled by the client libraries.

In general, I would expect the behavior of the migration to be that it _only_ adds batching, not that it re-enables queueing or retries.

Author: Prometheus exporter therefore has no reason to include QueueBatch settings in its call to initialize exporterhelper. If you don't want queueing, why include the queue sender? Also do you agree: if your answer is that you want splitting, not out-of-order dispatch, that num_consumers=1 should give you want you want?

Retries are _independent_ functionality in the exporterhelper: you can choose three separate parts of that library: the Timeout, the Retry, and the combined QueueBatch. Each of these can be selected/configured independently, so you can just avoid the parts you don't want.

Reviewer 2: >  If you don't want queueing, why include the queue sender? 

Because users will expect it and not necessarily understand that batching is not needed in this case (or at least that is what I argued for in open-telemetry/opentelemetry-collector-contrib/pull/42629 )

Reviewer 7: Sorry, I misremembered.  We don't use the *persistent* queue, and we can't use retries, but we do use the standard queue with num_consumers=1.

Author: I added a list of these reasons why an exporter will either not include queue/batch or will override the configuration.

Author: I listed the reasons you gave to motivate why exporters may opt-out.

callers (e.g., `otelarrowexporter`),

```go
func MigrateSpecifyQueueConfig(original queuebatch.Config) configoptional.Optional[queuebatch.Config] {
    // This will return the original configuration until the final
    // feature flag enabling wait_for_result for in-memory queues
    // with batching enabled by default. After this point, the original
    // configuration will be ignored.
}
```

Two feature flags have been introduced so far. They are logically
independent, but defined to advance in step with each other so that
users have independent control over `wait_for_result` and default
batching behavior. The feature flag state of these two will remain
equal.

> ## Full timeline

Reviewer 2: Can we summarize the expected user-facing behavior on each phase of a typical pipeline?

Author: Done, there's a new table at the beginning of the phases document.

The sequence of events is ordered in phases.

> ### Phase 1

Reviewer 2: I think we should have some way of notifying vendors about removing this on their official documentation on this phase. There are probably 100+ references to the batch processor on official vendors documentation:

<details>

<summary>Claude-generated summary of official documentation references to the batch processor</summary>

**I have only spot-checked the contents of this summary, they may be wrong, this is AI-generated**

### Alibaba Cloud — ARMS / Managed Service for OpenTelemetry

- **GitHub repo for issues:** *No public docs repo found.* Feedback is via the Alibaba Cloud console "Feedback" button on each doc page. The closest open-source repos are the demo collections at <https://github.com/alibabacloud-observability> (e.g. `alibabacloud-observability/java-demo`), but they do not host the docs.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OpenTelemetry integration guide (main) | <https://www.alibabacloud.com/help/en/arms/tracing-analysis/connect-opentelemetry-to-tracing-analysis> | Reference `opentelemetry-collector` ConfigMap uses `processors: batch: {}` alongside `memory_limiter` in trace, metric, and log pipelines. |
| 2 | Official docs — Use OpenTelemetry for tracing in Istio (ARMS) | <https://www.alibabacloud.com/help/en/arms/tracing-analysis/use-opentelemetry-to-perform-tracing-analysis-on-istio> | ConfigMap for the in-cluster Collector uses `processors: batch:` for logs and traces pipelines. |
| 3 | Official docs — Use OpenTelemetry for tracing in Istio (alt path) | <https://www.alibabacloud.com/help/en/opentelemetry/user-guide/use-opentelemetry-to-perform-tracing-analysis-on-istio> | Mirror of the Istio tracing page under the `opentelemetry/user-guide` hierarchy. |
| 4 | Official docs — Use OpenTelemetry to trace APISIX | <https://www.alibabacloud.com/help/en/arms/tracing-analysis/use-opentelemetry-to-trace-links-to-apisix> | APISIX → OTel Collector → ARMS path uses `processors: batch:` with the `otlphttp` exporter. |
| 5 | Official docs — Use OpenTelemetry to trace NGINX links | <https://www.alibabacloud.com/help/en/arms/tracing-analysis/use-opentelemetry-to-trace-nginx-links> | NGINX `ngx_otel_module` deployment instructions configure an OTel Collector with `processors: batch:`. |
| 6 | Official docs — Enable tracing for ACK data plane (Kubelet) | <https://www.alibabacloud.com/help/en/opentelemetry/user-guide/enable-tracing-for-the-ack-cluster-data-plane-kubelet-component> | DaemonSet `opentelemetry-collector-agent` configuration documents the `batch` processor in pipelines. |

---

### AWS — AWS Distro for OpenTelemetry (ADOT)

- **GitHub repo for issues:** <https://github.com/aws-observability/aws-otel-collector> (the distribution and its templates). The docs site `aws-otel.github.io` is itself generated from <https://github.com/aws-observability/aws-otel.github.io>.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Datadog partner exporter guide | <https://aws-otel.github.io/docs/partners/datadog/> | Recommends a `batch` processor with `timeout: 10s` as a required component for Datadog traces. |
| 2 | Official docs — Processors page | <https://aws-otel.github.io/docs/components/processors/> | Lists `batch` as a supported processor; describes ordering with tail sampling / group-by-trace. |
| 3 | Official docs — Advanced sampling guide | <https://aws-otel.github.io/docs/getting-started/advanced-sampling/> | Sample configuration includes `batch` processor with `send_batch_max_size: 8196`. |
| 4 | Official docs — ADOT EKS add-on collector configuration | <https://aws-otel.github.io/docs/getting-started/adot-eks-add-on/config-collector-intro/> | Reference configs (Prometheus / CloudWatch / X-Ray) include `batch`. |
| 5 | GitHub — distribution repo (templates & examples) | <https://github.com/aws-observability/aws-otel-collector> | Default and example configurations across the repo reference `batch`. |

---

### Chronosphere

- **GitHub repo for issues:** *No public docs repo found.* Feedback channel is `https://docs.chronosphere.io` → in-page feedback widget or via Chronosphere support.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OTel ingestion guide (metrics & traces) | <https://docs.chronosphere.io/ingest/metrics-traces/otel/otel-ingest> | "Configure batch processing" section instructs users to add the `batch` processor and shows the recommended pipeline `[transform, batch, resourcedetection, resource/service-instance]`. |
| 2 | Official docs — OTel ingestion guide (alt path) | <https://docs.chronosphere.io/ingest/otel/otel-ingest> | Parallel doc covering OTel ingestion; same `batch` guidance. |

---

### ClickHouse — ClickStack

- **GitHub repo for issues:** <https://github.com/ClickHouse/clickhouse-docs> (the docs site that powers `clickhouse.com/docs`; the README explicitly says "Issues with the documentation can be created in the ClickHouse/clickhouse-docs repo").
- Related upstream component: <https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/clickhouseexporter> — the ClickHouse exporter README literally states *"Always add batch-processor to collector pipeline"*. Worth raising in parallel; updating the upstream message will propagate to downstream vendor docs.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — ClickStack OTel collector ingestion | <https://clickhouse.com/docs/use-cases/observability/clickstack/ingesting-data/otel-collector> | Reference Collector config for ClickStack uses `processors: batch: timeout: 1s, send_batch_size: 10000` in logs / metrics / traces pipelines. |

---

### Coralogix

- **GitHub repo for issues:** <https://github.com/coralogix/telemetry-shippers> (the public Helm chart / distribution repo). Coralogix's main docs site (`coralogix.com/docs`) does not appear to have a public docs repo; doc feedback is typically via the chart repo issues.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Getting started with OpenTelemetry | <https://coralogix.com/docs/opentelemetry/getting-started/> | Explicitly recommends `batch: send_batch_size: 1024, send_batch_max_size: 2048, timeout: "1s"` for "all collectors". |
| 2 | Official docs — Install OTel on EC2 | <https://coralogix.com/docs/opentelemetry/configuration-options/install-opentelemetry-on-an-ec2-instance/> | Repeats the `batch` recommendation. |
| 3 | Official docs — Windows Event Logs & OpenTelemetry | <https://coralogix.com/docs/opentelemetry/configuration-options/windows-event-logs-and-opentelemetry/> | Sample `config.yaml` uses `batch` in a logs pipeline. |
| 4 | Official docs — Kubernetes Observability FAQs | <https://coralogix.com/docs/opentelemetry/kubernetes-observability/faqs/> | Recommends `batch` defaults for Coralogix's `otel-integration` Helm chart. |
| 5 | Official guide / tutorial | <https://coralogix.com/guides/opentelemetry/opentelemetry-collector-basics-tutorial/> | Names `batch` as a "built-in" processor. |
| 6 | GitHub — `otel-integration` Helm chart `values.yaml` | <https://github.com/coralogix/telemetry-shippers/blob/master/otel-integration/k8s-helm/values.yaml> | Chart enables `batch: enabled: true` by default with `sendBatchSize`, `sendBatchMaxSize`, `timeout` commented examples. |

---

### Dash0

- **GitHub repo for issues:** <https://github.com/dash0hq> is the public org; the integration hub pages are not in a public docs repo. Useful product repos include `dash0hq/dash0-operator` and `dash0hq/opentelemetry-collector-builder`. **Dash0 has published a blog post arguing that the batch processor should be retired** (<https://www.dash0.com/blog/why-the-opentelemetry-batch-processor-is-going-away-eventually>) — they are a likely ally for getting the integration page updated, even though the blog itself is excluded from this inventory.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OpenTelemetry Collector integration with Dash0 | <https://www.dash0.com/hub/integrations/int_opentelemetry-collector/overview> | Daemonset / deployment Helm value snippets configure the OTel Collector for Dash0 ingestion with `batch` in pipelines. |

---

### Datadog

- **GitHub repo for issues:** <https://github.com/DataDog/documentation> (the source for `docs.datadoghq.com`; most pages have an "Edit" link that goes to this repo and the README explicitly invites PRs).
- Related upstream component: <https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/datadogexporter> — the upstream example `collector.yaml` has already been migrated to `sending_queue.batch`, which makes a strong anchor for vendor-doc PRs.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Batch and Memory Settings (dedicated page) | <https://docs.datadoghq.com/opentelemetry/config/collector_batch_memory/> | Entire page is dedicated to configuring the `batch` processor with Datadog-specific values (`send_batch_max_size: 1000`, `send_batch_size: 100`, `timeout: 10s`). |
| 2 | Official docs — Install & configure the OTel Collector | <https://docs.datadoghq.com/opentelemetry/setup/collector_exporter/> | Setup overview links to `batch` configuration as a standard step. |
| 3 | Official docs — Setup the OTel Collector with the Datadog Exporter | <https://docs.datadoghq.com/opentelemetry/setup/collector_exporter/install/> | OOTB example references `batch` and points to `ootb-ec2.yaml` example in the contrib repo. |
| 4 | Official docs — Integrations tile for OpenTelemetry | <https://docs.datadoghq.com/integrations/otel/> | Minimal example uses `processors: batch: timeout: 10s`. |
| 5 | GitHub examples — Datadog exporter OOTB collector config | <https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/examples/collector.yaml> | The reference example file linked from Datadog docs. **Note:** this upstream file has already migrated to `sending_queue.batch`, but the Datadog vendor docs still reference the legacy `batch` processor — a useful inconsistency to cite when filing issues. |

---

### Dynatrace

- **GitHub repo for issues:** <https://github.com/Dynatrace/dynatrace-otel-collector> (the official Dynatrace distribution; config examples live in `config_examples/`). `docs.dynatrace.com` is not in a public GitHub repo, but the distribution repo is the channel maintainers monitor.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Batch OTLP requests use case | <https://docs.dynatrace.com/docs/ingest-from/opentelemetry/collector/use-cases/batch> | Dedicated use-case page recommending the `batch` processor for all setups. |
| 2 | Official docs — alternate URL of the same use case | <https://docs.dynatrace.com/docs/extend-dynatrace/opentelemetry/collector/use-cases/batch> | Mirror of the page above under the `extend-dynatrace` hierarchy. |
| 3 | Official docs — Configure the OTel Collector | <https://docs.dynatrace.com/docs/ingest-from/opentelemetry/collector/configuration> | Example YAML references `batch` in the pipelines section. |
| 4 | Official docs — OTel Collector overview | <https://docs.dynatrace.com/docs/ingest-from/opentelemetry/collector> | Lists `batch` as a recommended processor. |
| 5 | Official docs — Ingest NetFlow with OTel Collector | <https://docs.dynatrace.com/docs/ingest-from/opentelemetry/collector/use-cases/netflow> | NetFlow ingestion example uses `batch: send_batch_size: 30, timeout: 30s`. |
| 6 | Enablement Lab / training — Kubernetes OpenTelemetry | <https://dynatrace-wwse.github.io/enablement-kubernetes-opentelemetry/5-opentelemetry-traces/> | Training material instructs students to add `batch` to their pipelines. Backing repo: <https://github.com/dynatrace-wwse/enablement-kubernetes-opentelemetry>. |
| 7 | GitHub — Dynatrace OTel Collector distribution config examples | <https://github.com/Dynatrace/dynatrace-otel-collector/tree/main/config_examples> | Example YAMLs in the distribution repo include `batch` in pipelines. |

---

### Elastic

- **GitHub repo for issues:** <https://github.com/elastic/docs-content> (the current docs repo for v9.0+; the README invites issues and PRs). Legacy content lives in <https://github.com/elastic/observability-docs> and <https://github.com/elastic/stack-docs>.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Contrib OTel Collectors and language SDKs | <https://www.elastic.co/docs/solutions/observability/apm/opentelemetry/upstream-opentelemetry-collectors-language-sdks> | Recommends "the Batch processor and the memory limiter processor". |

---

### Google Cloud — Google-Built OpenTelemetry Collector / Ops Agent

- **GitHub repo for issues:** <https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector> (the distribution / Ops Agent collector). The Google Cloud docs site itself is not open source; the distribution repo is the place where defaults are generated by `distrogen` from `specs/google-built-opentelemetry-collector.yaml`.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Google-Built OTel Collector overview | <https://docs.cloud.google.com/stackdriver/docs/instrumentation/google-built-otel> | Describes the `batch` processor as a built-in component that "collects traces, metrics, or logs and bundles them into batches". |
| 2 | Official docs — Deploy on GKE | <https://docs.cloud.google.com/stackdriver/docs/instrumentation/opentelemetry-collector-gke> | Default GKE config places `batch` in the logs, metrics, and traces pipelines. |
| 3 | Official docs — Deploy on Compute Engine | <https://docs.cloud.google.com/stackdriver/docs/instrumentation/opentelemetry-collector-gce> | Default GCE config includes `batch`. |
| 4 | Official docs — Deploy on Cloud Run | <https://docs.cloud.google.com/stackdriver/docs/instrumentation/opentelemetry-collector-cloud-run> | Cloud Run sidecar config includes `batch` in all three pipelines. |
| 5 | Official docs — Deploy on Container-Optimized OS | <https://docs.cloud.google.com/stackdriver/docs/instrumentation/opentelemetry-collector-cos> | COS reference config includes `batch`. |
| 6 | Official docs — Managed Service for Prometheus setup (OTel) | <https://docs.cloud.google.com/stackdriver/docs/managed-prometheus/setup-otel> | Recommends `batch: send_batch_max_size: 200, send_batch_size: 200, timeout: 5s`. |

---

### Grafana Labs (Tempo, Loki, Alloy, Grafana Cloud)

- **GitHub repos for issues:** Grafana publishes its docs source code in product-specific repos, each with a `docs/` subtree:
  - General Grafana / website docs: <https://github.com/grafana/website>
  - Loki: <https://github.com/grafana/loki> (`docs/`)
  - Tempo: <https://github.com/grafana/tempo> (`docs/`)
  - Alloy: <https://github.com/grafana/alloy> (`docs/`)
  - Agent (legacy): <https://github.com/grafana/agent>

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Set up OTel Collector for Application Observability | <https://grafana.com/docs/opentelemetry/collector/opentelemetry-collector/> | Recommended manual `config.yaml` includes `batch` in traces / metrics / logs pipelines. |
| 2 | Official docs — Loki getting started with OTel Collector | <https://grafana.com/docs/loki/latest/send-data/otel/otel-collector-getting-started/> | Tutorial walks through configuring `processors: batch`. |
| 3 | Official docs — Enterprise Logs OTel Collector tutorial | <https://grafana.com/docs/enterprise-logs/latest/send-data/otel/otel-collector-getting-started/> | Same tutorial as above for Enterprise Logs. |
| 4 | Official docs — Grafana Alloy OTel to LGTM stack | <https://grafana.com/docs/alloy/latest/collect/opentelemetry-to-lgtm-stack/> | Uses `otelcol.processor.batch` (Alloy's wrapping of the same component). |
| 5 | Official docs — Grafana Agent Flow OTel to Grafana stack | <https://grafana.com/docs/agent/latest/flow/tasks/opentelemetry-to-lgtm-stack/> | Same as above for the legacy Agent Flow. |

---

### Honeycomb

- **GitHub repo for issues:** `docs.honeycomb.io` is hosted on Mintlify and **does not have a public source repo**. Each docs page has a "Report incorrect code" button (top right) — this is the official feedback channel. The closest public org is <https://github.com/honeycombio> (for product-specific issues, e.g. on `honeycombio/refinery`).

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Send data with the OTel Collector | <https://docs.honeycomb.io/send-data/opentelemetry/collector> | Full reference config "includes the batch processor and several extensions". |
| 2 | GitHub — Honeycomb metrics-flavored Collector configs | <https://github.com/honeycombio/opentelemetry-collector-configs> | Repo of recommended Honeycomb collector configurations; pipelines reference `batch`. |

---

### IBM Instana

- **GitHub repo for issues:** <https://github.com/instana/instana-otel-collector> (the official Instana OTel distribution).

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | GitHub — Instana Distribution of OpenTelemetry Collector | <https://github.com/instana/instana-otel-collector> | Distribution README and configuration examples use `batch`. |
| 2 | Training — WAIOps Tech Jam OTel Collector lab | <https://ibm.github.io/waiops-tech-jam/labs/instana/opentelemetry/otel-collector/> | Hands-on lab teaches students to configure the OTel Collector for Instana including `batch`. Backing repo: <https://github.com/IBM/waiops-tech-jam>. |

---

### Lightstep / ServiceNow Cloud Observability

- **GitHub repos for issues:**
  - <https://github.com/lightstep/sn-collector> (the ServiceNow distribution of the OTel Collector)
  - <https://github.com/lightstep/otel-collector-charts> (the Helm charts, including `kube-otel-stack`)

  The `docs.lightstep.com` site is not in a public repo.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Recommended Collector configuration | <https://docs.lightstep.com/docs/recommended-collector-config> | Reference config recommends `batch` processor as a baseline. |
| 2 | Official docs — Use the OpenTelemetry Collector | <https://docs.lightstep.com/docs/collector-home-page> | Overview that references `batch` as a recommended processor. |
| 3 | Official docs — Performance test and tune the Collector | <https://docs.lightstep.com/docs/otel-collector-prometheus-receiver-performance-tuning> | "The batch processor is also highly recommended"; tuning guidance on `send_batch_size`, `timeout`, `send_batch_max_size`. |
| 4 | Training / workshop | <https://workshops.lightstep.com/workshops/nodejs-logs/container-logs/> | Workshop pipeline uses `processors: [batch]` for logs. |
| 5 | GitHub — Lightstep `otel-collector-charts` (kube-otel-stack) | <https://github.com/lightstep/otel-collector-charts/blob/main/charts/kube-otel-stack/values.yaml> | Helm chart pipelines include `batch`. |
| 6 | GitHub — ServiceNow distribution `sn-collector` | <https://github.com/lightstep/sn-collector> | Distribution README lists `concurrentbatch` (in development) as a successor; current docs and dependencies still reference the legacy `batch` processor in many examples. |

---

### Logz.io

- **GitHub repo for issues:** <https://github.com/logzio/opentelemetry-examples> (examples linked from the docs). The `docs.logz.io` source repo is not public.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Ship OpenTelemetry data | <https://docs.logz.io/docs/shipping/other/opentelemetry-data/> | Configuration walkthrough adds the `batch` processor to traces, metrics, and logs pipelines. |

---

### Microsoft Azure — Azure Monitor / Application Insights

- **GitHub repo for issues:** <https://github.com/MicrosoftDocs/azure-docs> (the source for `learn.microsoft.com/azure/...`; every Azure docs page has an "Edit" pencil that opens a PR against this repo).

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OTel in Application Insights | <https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-configuration> | Code samples include `processors: batch` when chaining the OTLP exporter from a Collector to Application Insights. (Primarily SDK-side `BatchSpanProcessor`, but Collector pipeline examples reference the `batch` processor too.) |

---

### New Relic

- **GitHub repo for issues:** <https://github.com/newrelic/docs-website> (the source for `docs.newrelic.com`; every page has an "Edit" button → GitHub).
- Related distribution: <https://github.com/newrelic/opentelemetry-collector-releases>.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OTel Collector for data processing | <https://docs.newrelic.com/docs/opentelemetry/get-started/collector-processing/opentelemetry-collector-processing-intro/> | Dedicated "Batch processor" section instructs users to "configure the Collector to use the batch processor". Minimal example uses `processors: batch` in all three pipelines. |
| 2 | Official docs — Legacy URL of the same page | <https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/opentelemetry-collector-intro> | Older path that still redirects / serves the same content. |
| 3 | GitHub — `newrelic-opentelemetry-examples` | <https://github.com/newrelic/newrelic-opentelemetry-examples/tree/main/other-examples/collector/nr-config> | Repository linked from official docs as the "working example"; pipelines use `batch`. |

---

### OpenObserve

- **GitHub repo for issues:** <https://github.com/openobserve/openobserve-helm-chart> (the chart whose generated default Collector config contains `batch`). Main product repo: <https://github.com/openobserve/openobserve>.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | GitHub — `openobserve-helm-chart` collector default | <https://github.com/openobserve/openobserve-helm-chart/issues/117> | Upstream-generated config produced by the OpenObserve collector chart includes `processors: batch: {}` in pipelines. |

---

### SigNoz

- **GitHub repos for issues:**
  - <https://github.com/SigNoz/signoz.io> (the docs site source)
  - <https://github.com/SigNoz/signoz> (main product repo; collector default configs live here)

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OTel Collector configuration | <https://signoz.io/docs/opentelemetry-collection-agents/opentelemetry-collector/configuration/> | Reference pipeline lists `batch` as a default and recommends placing it last. |
| 2 | Official docs — VM collection agent install | <https://signoz.io/docs/opentelemetry-collection-agents/vm/install/> | Tailored `config.yaml` uses `batch: send_batch_size: 1000, timeout: 10s`. |
| 3 | Official docs — Docker collection agent configure | <https://signoz.io/docs/opentelemetry-collection-agents/docker/configure/> | Recommends `batch: send_batch_size: 1000, send_batch_max_size: 2048, timeout: 10s`. |
| 4 | Official docs — Docker Swarm collection agent configure | <https://signoz.io/docs/opentelemetry-collection-agents/docker-swarm/configure/> | Same `batch` recommendation. |
| 5 | Official docs — Host metrics with OpenTelemetry | <https://signoz.io/docs/infrastructure-monitoring/hostmetrics/> | Host-metrics pipeline uses `processors: [resourcedetection, batch]`. |
| 6 | Official docs — GeoIP processor guide | <https://signoz.io/docs/opentelemetry-collection-agents/opentelemetry-collector/geoip-processor/> | Recommends `[memory_limiter, geoip, batch]` pipeline pattern. |

---

### Splunk Observability Cloud (and AppDynamics)

- **GitHub repos for issues:**
  - <https://github.com/signalfx/splunk-otel-collector> (the official Splunk distribution; defaults defined here)
  - <https://github.com/signalfx/splunk-otel-collector-chart> (the Helm chart, including advanced configuration docs in `docs/advanced-configuration.md`)
  - `docs.splunk.com` / `help.splunk.com` are not in a public docs repo; the distribution repo is the path of least resistance.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Batch processor (dedicated page) | <https://docs.splunk.com/observability/gdi/opentelemetry/components/batch-processor.html> | Dedicated reference page; states that `batch` is part of the **default configuration** of the Splunk Distribution of the OTel Collector in agent and gateway modes. |
| 2 | Official docs — same page (new help URL) | <https://help.splunk.com/en/splunk-observability-cloud/manage-data/splunk-distribution-of-the-opentelemetry-collector/get-started-with-the-splunk-distribution-of-the-opentelemetry-collector/collector-components/processors/batch-processor> | Identical content under `help.splunk.com`. |
| 3 | Official docs — Get started with Splunk Distribution of OTel Collector | <https://help.splunk.com/en/splunk-observability-cloud/manage-data/splunk-distribution-of-the-opentelemetry-collector/get-started-with-the-splunk-distribution-of-the-opentelemetry-collector> | Pipeline examples include `batch`. |
| 4 | Official docs — Configure Collector tutorial (Linux) | <https://docs.splunk.com/observability/en/gdi/opentelemetry/collector-linux/collector-configuration-tutorial/collector-config-tutorial-edit.html> | Tutorial keeps `batch` in logs pipeline. |
| 5 | Official docs — Same tutorial (`help.splunk.com`) | <https://help.splunk.com/en/splunk-observability-cloud/get-started/examples-and-tutorials/splunk-observability-cloud-tutorials/opentelemetry-collector-tutorials/configure-the-splunk-distribution-of-opentelemetry-collector-on-a-linux-host/part-2-edit-the-collector-configuration-to-add-new-components> | Same content under `help.splunk.com`. |
| 6 | Official docs — AppDynamics SaaS, Configure Processors | <https://help.splunk.com/en/appdynamics-saas/application-performance-monitoring/25.5.0/splunk-appdynamics-for-opentelemetry/configure-the-opentelemetry-collector/configure-processors> | Recommends `batch` defaults of `timeout: 30s`, `send_batch_size: 8192`. |

---

### Sumo Logic

- **GitHub repo for issues:** <https://github.com/SumoLogic/sumologic-otel-collector> (the official Sumo Logic distribution; `docs/configuration.md` holds the examples). `www.sumologic.com/help/docs` is not in a public repo.

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — Sumo Logic OTel Collector overview | <https://www.sumologic.com/help/docs/send-data/opentelemetry-collector/> | Lists `batch` among the processors used "in every pipeline". |
| 2 | Official docs — Additional configurations reference | <https://www.sumologic.com/help/docs/send-data/opentelemetry-collector/data-source-configurations/additional-configurations-reference/> | Recommends `batch: send_batch_size: 1024, timeout: 1s, send_batch_max_size: 2048`. |
| 3 | Official docs — Data source configurations overview | <https://www.sumologic.com/help/docs/send-data/opentelemetry-collector/data-source-configurations/overview/> | Reference covers the role of `batch`. |
| 4 | Official docs — Set up Traces Collection for AWS environments | <https://www.sumologic.com/help/docs/apm/traces/get-started-transaction-tracing/set-up-traces-collection-aws-environments/> | Default config uses two named instances: `batch/traces` and `batch/metrics`. |
| 5 | Official docs — Archive log data to S3 with OTel | <https://www.sumologic.com/help/docs/manage/data-archiving/archive-otel/> | Archive feature explicitly relies on the OTel `batch` processor to control archive file size/window. |
| 6 | Official docs — OTel Collector Insights app | <https://www.sumologic.com/help/docs/integrations/sumo-apps/opentelemetry-collector-insights/> | Dashboards include "batch processing efficiency and triggers" — built around `otelcol_processor_batch_*` metrics. |
| 7 | GitHub — `sumologic-otel-collector` docs | <https://github.com/SumoLogic/sumologic-otel-collector/blob/main/docs/configuration.md> | Distribution configuration docs include `batch` examples. |

---

### Uptrace

- **GitHub repo for issues:** <https://github.com/uptrace/uptrace> (the main product repo; docs source is in `web/` / `docs/`).

| # | Type of documentation | URL | Notes |
|---|-----------------------|-----|-------|
| 1 | Official docs — OTel Collector configuration tutorial | <https://uptrace.dev/opentelemetry/collector/config> | Tutorial includes the `batch` processor. |
| 2 | Official docs — OTel Collector exporters guide | <https://uptrace.dev/opentelemetry/collector/exporters> | Sample pipelines for traces / metrics / logs use `processors: [batch]` with the Uptrace OTLP exporter. |
| 3 | Official docs — Getting started with the OTel Collector | <https://uptrace.dev/opentelemetry/collector> | Overview recommends `batch` as one of the main reasons to deploy a Collector. |

</details>

Author: Yes! I think a blog post would help spread the word.

Author: I have described a blog post that would go out around v0.158 to notify everyone.

These are loosely dependent,

1. Define `StandardQueueConfig()` function.
2. Ensure all exporters in OpenTelemetry repos define `queue_sender` having

>    type `configoptional.Optional[exporterhelper.QueueBatchConfig]`

Reviewer 5: How? Is this manual, or should we introduce an mdatagen-generated test that ensures the default config contains `queue_sender` with the standard config by default? Exporters could explicitly opt out during the rollout in metadata.yaml.

> 3. Define the two feature flags.

Reviewer 2: ```suggestion
3. Introduce the two feature gates in alpha stability.
```

Author: I updated to be clear about the feature gate level, thanks.

4. Support a configurable metrics prefix to distinguish processor
   batching from exporter batching
   ([#14038](https://github.com/open-telemetry/opentelemetry-collector/issues/14038)).
5. Implement `queuebatchprocessor`
   ([#13583](https://github.com/open-telemetry/opentelemetry-collector/pull/13583)).
6. Implement `batchmigration` and two migration functions for
   non-standard callers.

### Phase 2

7. Deprecate `batchprocessor`.
8. Full audit of exporters and default settings. Exporters can opt-out
   of the migration here, replace callsites use of `configoptional`

>    with one of `StandardQueueConfig`, `MigrateQueueConfig`, and

Reviewer 5: I'm not sure I understand the `batchmigration` part. Here's my partial understanding:
 - `StandardQueueConfig` is clear, it means that an exporter is opting into the new standard default.
 - `MigrateSpecifyQueueConfig` means that the exporter is opting out, and its own supplied default will be used until phase 3 (or unless the exporterQueueBatchEnabled feature gate is enabled?)
 - `MigrateQueueConfig`... I don't know what this does

   `MigrateSpecifyQueueConfig`.

> 9. Implement coordinated feature behavior. The default changes

Reviewer 2: Does this mean we move the feature gates to beta?

Author: Yes, thanks. I added more precision about the feature flag timeline.

   recommended above will be applied through the three functions named
   above. The batch processor will be programmed to act as a
   pass-through when the `exporterhelper.exporterQueueBatchEnabled` is
   enabled.

> After the batch processor is deprecated, we will wait 6 months for

Reviewer 8: With a constant and predictable release cycle of two weeks, should a concret version be named? 6 months can mean different targets, depending on when people read it.

Author: Yes. I have tentatively named 0.158 as the point where phase 2 begins, with subsequent points 0.164 and 0.170.

> After the batch processor is deprecated, we will wait 6 months for

Reviewer 2: It would be very useful to have a list of adopters of the feature gates that can provide feedback for moving forward to gain confidence on the feature gates working as expected. could we have that?

Author: Added.

users to choose a migration path through configuration.

### Phase 3

10. Update various documentation pointing to `queuebatchprocessor`,
    which will advise users to configure exporter batching in most
    cases. ([#13766](https://github.com/open-telemetry/opentelemetry-collector/issues/13766))
11. After waiting the specified period, we will change both feature
    flags to Stable. At this time, the behavior of
    `StandardQueueConfig`, `MigrateQueueConfig`,
    `MigrateSpecifyQueueConfig`, and `NewDefaultQueueConfig()` will
    change while `batchprocessor` is simultaneously disabled. All
    exporters have `wait_for_result: true` by default, which is
    overridden by defining a storage extension.

> 12. Update standard Helm charts. Simply remove the batch processor,

Reviewer 2: Why not update this in phase 2?

Author: I changed this to happen towards the end of phase 2. As recommended, I think we should have large-scale users agree that the feature flag works before we move to phase 3, this is probably also a good precondition for updating the default Helm configuration.

Author: If we do this at the start of phase 2, it's risky. I think it can be done towards the end of phase 2. I have described exit criteria for phase 2, suggesting we gate helm changes on the same criteria.

    because exporters batch by default excepting those that opt out.
13. Deprecate `MigrateQueueConfig`, `MigrateSpecifyQueueConfig`, and
    the `batchmigration` package. Users of the migration helpers
    will switch to `StandardQueueConfig` in most cases.

### Phase 4

We will wait another 6 months before completing these steps.

> 14. Remove the `batchprocessor` from the core distribution.

Reviewer 2: Will we ever remove it from the core repository?

Author: This is what I meant. We actually remove the original component. We would introduce the `queuebatchprocessor` to replace it 6-12 months earlier.

Reviewer 2: The current wording suggested to me that you were talking about removing this line https://github.com/open-telemetry/opentelemetry-collector-releases/blob/e8f3db9c2f7ddfb21c5e84c9f295d0a820318e04/distributions/otelcol/manifest.yaml#L35 and not removing the module itself. I think we can separate the two steps if we want to, not necessarily something to decide right now

Author: Which do you recommend? I would remove the component and the manifest entry. 
I see how we could use this to lengthen the release timeline, for more safety.
At Phase 3, instead of removing the component we remove the manifest entry, that lets users keep the batch processor if they build it themselves. Then, the new warning requested by reviewers (combining batchprocessor and exporterhelper batching) will be meaningful, warning users that have released it themselves of the misconfiguration.

Even so, we believe many exporters will choose to enable batching by
default now that the feature is stable. This RFC calls for an audit of
the components, evaluating for every exporter whether a `Default` or
`None` should become `Some` in each exporter's default queue config.

A feature flag for this change will be defined,
`exporterhelper.exporterQueueBatchEnabled`.

#### Recommended final default queue configuration

The `NewDefaultQueueConfig` function will begin being influenced by
two feature flags.

```go
func NewDefaultQueueConfig() queuebatch.Config {
	return queuebatch.Config{
		Sizer:           request.SizerTypeRequests,
		NumConsumers:    10,
		QueueSize:       1_000,
        // CHANGED: Blocking is becoming enabled by default.
		BlockOnOverflow: exporterQueueBlockOnOverflow.IsEnabled(),
        WaitForResult:   false,
		Batch: configoptional.DefaultOrSome(
            // CHANGED: Batching is becoming enabled by default.
            exporterQueueBatchEnabled.IsEnabled(),
            queuebatch.BatchConfig{
			    FlushTimeout: 200 * time.Millisecond,
			    Sizer:        request.SizerTypeItems,
			    MinSize:      8192,
            }),
	}
}
```

> The above makes use of an hypothetical `configoptional` helper:

Reviewer 5: Nit: not convinced this helper is worthwhile, could be inlined in NewDefaultQueueConfig.
Has a need for this come up elsewhere?

Author: It's not needed. I think it makes the document more readable.

   not yet migrated. Shipping the updated chart is the signal that we
   are ready to enter Phase 3.

### Phase 3

In a single release cycle:

11. Remove `batchprocessor` from the standard OpenTelemetry Collector
    distribution manifest. The component's Go module remains in the
    core repository so that custom builds (e.g., via `ocb`) can
    continue to include it as an explicit dependency. The startup
    warning from Phase 1, step 5 keeps flagging misconfigurations in
    such custom builds.
12. Promote the feature gate to **Beta** (default on). Exporterhelper
    enables batching by default; users can still opt out via the gate
    for one cycle.

### Phase 4

We will wait another +6 releases (~3 months).

13. Promote the feature gate to **Stable** and remove it. The new
    exporterhelper batching default becomes permanent and can no
    longer be toggled via the gate.
14. Remove the `batchprocessor` Go module from the core repository.
    Custom builds that still reference it must move to
    `queuebatchprocessor` or vendor the component themselves.

## Additional work items

Items noted in this document that should be tracked separately,
accomplished during Phase 1:

> - Reject `wait_for_result: true` combined with a configured storage

Reviewer 9: This check already exist I guess. We can remove this here since we have the expectation sentenced above under `Recommendation: no
  error propagation by default`

https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/internal/queuebatch/config.go#L77

Author: True. I will follow up.

exporterhelper defaults, users will pay the cost of batching and then
re-batching.

Users have 6 releases (~3 months) during which the batch processor
will print a deprecation warning. During this window the new
exporterhelper defaults are still off, so unmodified pipelines do not
double-batch. Adopters who manually enable the Beta-on values to
provide feedback, and custom builds that continue to ship
`batchprocessor` after it is dropped from the default distribution in
Phase 3, are protected by the startup warning introduced in Phase 1,
step 5.

## Risks

The phased timeline below is designed around the following risks:

- **Custom exporters that do not use exporterhelper lose batching
  entirely** once `batchprocessor` is removed. The Phase 2 deprecation
  window gives these users time to adopt `queuebatchprocessor` or move
  to exporterhelper.
- **`block_on_overflow: true` changes failure mode.** Pipelines that
  previously dropped silently on overflow will apply backpressure;
  receivers with timeouts may begin returning errors upstream. Must be

>   highlighted in release notes.

Reviewer 6: Maybe this has been brought up in previous threads, sorry if that's the case. But I am not convinced `block_on_overflow: true` is the right default.

`block_on_overflow: false` does not "drop silently" on overflow: it returns an error, which will propagate back to the receiver, and then to the client. This is the "error" style of backpressure, where we refuse to handle a payload when overloaded and rely on the client to retry later, like the `memory_limiter` processor does. (Unless there is another async component in the way, which I think should have retry logic on its own.)

`block_on_overflow: true` instead uses the "blocking" style of backpressure, which doesn't rely on retry logic upstream, but also has the major (in my opinion and limited experience) problem that it completely fails to cap the memory use of the Collector: instead of freeing up memory and letting the client store the payload until the next retry (which it would anyway), data starts accumulating in the form of blocked goroutines (assuming an OTLP-style receiver which spawns a Goroutine for each request). This is less efficient than accumulating data in the in-memory queue and just makes overload issues worse.

If we want a true drop-in replacement of the batch processor with all its quirks, maybe we should make `wait_for_result` or `block_on_overflow` the default, but _only_ in the new `queue_batch_processor`?

Author: I have considered that `queuebatchprocessor` should use defaults as you suggest. I agree it should set `wait_for_result: true` and `block_on_overflow: true`.

I consider the failure scenario when `block_on_overflow: false` to be a bad outcome. AFAICT the OTLP receiver does not know to send a "RetryAfter" header using the failure returned by the queue, which means the most likely outcome is that the exporter re-sends the same request as a result of its own retry logic. So, in an attempt to save memory, we end up with a failure and the same request sent again, causing more garbage collection than if the original request had been allowed to wait. OTOH if we knew that (a) the exporter respects RetryAfter and (b) the receiver propagates RetryAfter, then I can be convinced that fail-fast is smart. However, I'd like to see evidence that real OTel SDKs actually do this, moreover I think we should have evidence that SDKs are inclined to reconnect to a new backend when there is a load-balanced endpoint and they receive a RetryAfter response.

Memory limiter extension (no one should use the processor, IMO) also supports blocking instead of failing for the same reason: it is wasteful to return failure when there is memory available. The reason I prefer blocking is that when we block, we cause the sender to slow down. They are forced to drop their own data because their queues fill, and as a result the system does less work overall under the overload.

Reviewer 6: > causing more garbage collection than if the original request had been allowed to wait

I think the amount of CPU caused by having more memory to free is minor compared to the amount of CPU spent tracing through many accumulating goroutines blocked by a full queue. In my limited experience running struggling Collectors `block_on_overflow: true` is really bad for recovery from near-OOM situations.

> AFAICT the OTLP receiver does not know to send a "RetryAfter" header using the failure returned by the queue, which means the most likely outcome is that the exporter re-sends the same request as a result of its own retry logic.

Do we absolutely need a RetryAfter header? I think even simple exponential backoff (which I would assume most OTLP exporters already have) can help relieve memory pressure a bit. And a common alternative to what you're calling "fail fast" is "fail hard", ie. the Collector OOMs and all in-memory data is lost, which I think tilts the balance. The only way to avoid that is using the memory limiter processor, which has the same issue with quick retries.

Author: Discussed in the 6/3/26 SIG. I will stop proposing to change block_on_overflow for the exporterhelper.

I will propose that queuebatchprocessor adopts both `block_on_overflow: true` and `wait_for_result: true` to retain the original processor's blocking behavior.

Author: I have done this. https://github.com/open-telemetry/opentelemetry-collector/pull/15273/changes/7b8749e2f011215c4848abb066a8195bbb1659a2
